src/parser/mod.rs

```diff
@@ -11491,7 +11495,14 @@ impl<'a> Parser<'a> {
 
     /// Parse a TOP clause, MSSQL equivalent of LIMIT,
     /// that follows after `SELECT [DISTINCT]`.
-    pub fn parse_top(&mut self) -> Result<Top, ParserError> {
+    pub fn maybe_parse_top(
+        &mut self,
+        is_before_distinct: bool,
+    ) -> Result<Option<Top>, ParserError> {
```

Comment:
i'm thinking this this function probably wont need to change, for the 'maybe' part, the caller can check the next token before calling
```rust
if self.expects_top_before_distinct() && self.peek_token() == TOP
```

Then for the `is_before_distinct`, that flag I think would make sense as a `top_before_distinct` on the `SELECT` struct instead? Its a bit unusual that the top struct contains metadata about its positioning relative to another struct given that both structs are not necessarily always tied to each other. The enclosing SELECT on the other hand can contain this info since it gets to format both top and distinct in this case

Reply:
Putting the positioning hint on the `SELECT` makes sense I was initially intimidated by introducing a change to such a central (the most central?) struct :-) See next commit